fix: better cancelled jobs handling

rgrinberg opened the pull request:
* do not print output of cancelled jobs
* `Process.run` will not return with a value signalling success if the job was cancelled after the process terminated.

Should help out with #5482.

rgrinberg commented:
ping on this fix @snowleopard. It fixes quite an annoying issue.

snowleopard commented on src/dune_engine/process.ml:
```diff
@@ -801,20 +801,23 @@ let run_internal ?dir ?(stdout_to = Io.stdout) ?(stderr_to = Io.stderr)
-      in
-      (res, times))
+      if Fiber.Cancel.fired cancel then
+        raise Dune_util.Report_error.Already_reported
```
I find this line confusing, perhaps due to the ambiguous name `Already_reported`.

Here you are not reporting an error that wasn't reported before. Maybe some other error was "already reported" but not the one that we have here. 

Can we make the code/naming somehow less confusing?

snowleopard replied:
And, by the way, what if the job actually succeeded? Basically, the code seems to do the opposite of what you wrote in the PR description.

rgrinberg replied:
> And, by the way, what if the job actually succeeded? Basically, the code seems to do the opposite of what you wrote in the PR description.

Could you clarify this a bit?

So what if the job succeeded? if the build job it belonged to was cancelled, we aren't interested in its output.

snowleopard replied:
I think I basically don't understand what your second bullet point means. At least, I don't see any change that corresponds to it?

snowleopard replied:
Oh, sorry, I think I get it now. Ignore me, I'm slow.

rgrinberg replied:
Actually my description was hastily written. I'll update it and the commit message.

snowleopard replied:
So, this `if` that you added achieves two things at the same time and it wasn't obviuos to me. I suggest adding a comment that clarifies that.

rgrinberg replied:
> Can we make the code/naming somehow less confusing?

I suppose we abuse this name in other places as well. For example, we treat user requested shutdowns initiated with a signal as `Already_reported` as well. I'm at a loss for a better name right now, but I'll leave a comment to clarify.

snowleopard commented on src/dune_engine/scheduler.mli:
```diff
@@ -114,7 +114,7 @@ val t : unit -> t Fiber.t
 (** [with_job_slot f] waits for one job slot (as per [-j <jobs] to become
     available and then calls [f]. *)
-val with_job_slot : (Config.t -> 'a Fiber.t) -> 'a Fiber.t
+val with_job_slot : (Fiber.Cancel.t -> Config.t -> 'a Fiber.t) -> 'a Fiber.t
```
Please update the comment.

snowleopard reviewed (changes requested):
I left some questions/comments in the code. 

The main issue is that it seems like the change you are making doesn't achieve the second bullet point from the PR description?

rgrinberg commented:
I clarified the commit and added some comments. I also added a CHANGES entry that claims we're fixing #5482. Perhaps a bit optimistic, but we can revisit if the problem comes back.

snowleopard reviewed (approved):
Looks good now, thanks for the comments!

snowleopard commented:
> I clarified the commit and added some comments. I also added a CHANGES entry that claims we're fixing #5482. Perhaps a bit optimistic, but we can revisit if the problem comes back.

Agreed, it does look like this PR can fix it.
